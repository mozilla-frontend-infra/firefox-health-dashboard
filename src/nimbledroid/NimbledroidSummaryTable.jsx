import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import { generateSitesTableContent } from './config';
import fetchNimbledroidData from './NimbledroidApiHandler';
import { withErrorBoundary } from '../vendor/errors';
import SummaryTable from './SummaryTable';
import StatusWidget from './StatusWidget';

const styles = {
  root: {
    margin: '0 0 0.7rem 0',
  },
  summary: {
    color: 'black',
    width: '400px',
  },
};

class NimbledroidSummaryTable extends Component {
  state = {
    tableContent: null,
    showSites: false,
  };

  constructor(props) {
    super(props);
    const { configuration, nimbledroidData } = this.props;

    if (nimbledroidData) {
      this.state = generateSitesTableContent(nimbledroidData, configuration);
    }
  }

  async componentDidMount() {
    const { configuration } = this.props;
    const nimbledroidData = await fetchNimbledroidData(
      configuration.summaryTable,
    );

    this.setState(generateSitesTableContent(nimbledroidData, configuration));
  }

  render() {
    const {
      showSites, summary, tableContent, tableHeader,
    } = this.state;
    const { classes } = this.props;

    if (!tableContent) {
      return null;
    }

    return (
      <div className={classes.root}>
        {!showSites && (
          <div>
            <Button onClick={() => this.setState({ showSites: true })}>
              Show detailed view
            </Button>
            <div className={classes.summary}>
              {summary.map((s) => (
                <StatusWidget key={s.title.text} {...s} />
              ))}
            </div>
          </div>
        )}
        {showSites && (
          <div>
            <Button onClick={() => this.setState({ showSites: false })}>
              Show summary view
            </Button>
            <SummaryTable header={tableHeader} content={tableContent} />
          </div>
        )}
      </div>
    );
  }
}

NimbledroidSummaryTable.propTypes = {
  classes: PropTypes.shape({}),
  nimbledroidData: PropTypes.shape({}),
  configuration: PropTypes.shape({
    baseProduct: PropTypes.string.isRequired,
    compareProduct: PropTypes.string.isRequired,
    products: PropTypes.arrayOf(PropTypes.string).isRequired,
    targetRatio: PropTypes.number.isRequired,
  }).isRequired,
};

export default withStyles(styles)(withErrorBoundary(NimbledroidSummaryTable));
