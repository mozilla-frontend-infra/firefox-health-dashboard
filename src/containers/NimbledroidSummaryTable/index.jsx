import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import { generateSitesTableContent } from '../../utils/nimbledroid';
import fetchNimbledroidData from '../../utils/nimbledroid/fetchNimbledroidData';
import withErrorBoundary from '../../hocs/withErrorBoundary';

import SummaryTable from '../../components/SummaryTable';
import StatusWidget from '../../components/StatusWidget';

const styles = {
  root: {},
  summary: {
    color: 'black',
    width: '400px',
  },
};

class NimbledroidSummaryTable extends Component {
  state = {
    tableContent: null,
    showSites: false,
  };

  async componentDidMount() {
    const { configuration, nimbledroidData, handleError } = this.props;
    const { products } = configuration;

    try {
      const data = nimbledroidData || await fetchNimbledroidData(products);
      this.setTableContents(data, configuration);
    } catch (error) {
      handleError(error);
    }
  }

  setTableContents(nimbledroidData, configuration) {
    this.setState({
      ...generateSitesTableContent(nimbledroidData, configuration),
    });
  }

  render() {
    const { showSites, summary, tableContent, tableHeader } = this.state;
    const { classes } = this.props;

    if (!tableContent) {
      return null;
    }

    return (
      <div className={classes.root}>
        {!showSites && (
          <div>
            <Button onClick={() => this.setState({ showSites: true })}>
              Show detailed view
            </Button>
            <div className={classes.summary}>
              {summary.map(s => (<StatusWidget key={s.title.text} {...s} />))}
            </div>
          </div>
        )}
        {showSites && (
          <div>
            <Button onClick={() => this.setState({ showSites: false })}>
              Show summary view
            </Button>
            <SummaryTable
              header={tableHeader}
              content={tableContent}
            />
          </div>
        )}
      </div>
    );
  }
}

NimbledroidSummaryTable.propTypes = {
  classes: PropTypes.shape({}),
  nimbledroidData: PropTypes.shape({}),
  configuration: PropTypes.shape({
    baseProduct: PropTypes.string.isRequired,
    compareProduct: PropTypes.string.isRequired,
    products: PropTypes.arrayOf(PropTypes.string).isRequired,
    targetRatio: PropTypes.number.isRequired,
  }),
};

export default withErrorBoundary(withStyles(styles)(NimbledroidSummaryTable));
