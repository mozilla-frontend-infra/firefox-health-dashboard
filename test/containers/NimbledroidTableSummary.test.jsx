import React from 'react';
import renderer from 'react-test-renderer';
import NimbledroidSummaryTable from '../../src/containers/NimbledroidSummaryTable';

const nimbledroidData = require('../mocks/nimbledroidData.json');

it('renders correctly', () => {
  const tree = renderer
    .create((
      <NimbledroidSummaryTable
        nimbledroidData={nimbledroidData}
        configuration={{
          baseProduct: 'org.mozilla.focus',
          compareProduct: 'com.chrome.beta',
          products: [
            'org.mozilla.klar',
            'org.mozilla.focus',
            'com.chrome.beta',
          ],
          targetRatio: 1.2,
        }}
      />
    ))
    .toJSON();
  expect(tree).toMatchSnapshot();
});
