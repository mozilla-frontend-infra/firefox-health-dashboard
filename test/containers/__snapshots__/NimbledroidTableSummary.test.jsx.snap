// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="NimbledroidSummaryTable-root-1"
>
  <div>
    <button
      className="MuiButtonBase-root-29 MuiButton-root-3 MuiButton-text-5 MuiButton-flat-8"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-4"
      >
        Show detailed view
      </span>
      <span
        className="MuiTouchRipple-root-32"
      />
    </button>
    <div
      className="NimbledroidSummaryTable-summary-2"
    >
      <div>
        <div
          className="status-widget status-red padded"
        >
          <span
            className=""
          >
            GeckoView &gt; 20% slower than Chrome Beta
          </span>
          <span>
            37/46
          </span>
        </div>
      </div>
      <div>
        <div
          className="status-widget status-yellow padded"
        >
          <span
            className=""
          >
            GeckoView &gt; 0% and &lt;= 20% slower than Chrome Beta
          </span>
          <span>
            2/46
          </span>
        </div>
      </div>
      <div>
        <div
          className="status-widget status-green padded"
        >
          <span
            className=""
          >
            GeckoView &lt;= 0% (i.e. faster than) Chrome Beta
          </span>
          <span>
            7/46
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
